streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -160,11 +160,18 @@ public void processInOrder(final K key, final V value, final long timestamp) {
 
                     if (endTime < timestamp) {
                         leftWinAgg = next.value;
+                        // store the combined window if it is found so that a right window can be created for
+                        // the combined window's max record, as needed
                         if (isLeftWindow(next) || endTime == windows.timeDifferenceMs()) {
```

Comment:
Not sure I understand the `|| endTime == windows.timeDifferenceMs()`, is that left over from the 1st PR? It seems important to enforce that the window `latestLeftTypeWindow` points to is actually a left window

Reply:
It is, until we have a combined window that holds records that don't have corresponding left windows. I treated the combined window as a defacto left window since it's taking the place of early record's left windows. EX: for `timeDifference=10`, a record at 4 would be stored in [0,10]. If a record comes in at 11, we need the window from [5,15], but we only create that if we find a `leftTypeWindow`, which doesn't exist for 4. For these records, [0,10] is the only window we have, and therefore the closest thing to a left type window.

I didn't change variable names because I wanted the original algorithm to be understandable. It feels like adding early records has added a significant amount of complexity and I do wonder if it's worth it to have the addition confusion for testing flexibility - WDYT?

Reply:
I think it's way too confusing if `latestLeftTypeWindow` is not necessarily a left type window, we should definitely rename it. But I think we can pick a reasonable name -- the point of keeping this window is really just to learn the timestamp of the previous record, right? Do we even need the aggregate or any other data for this window, or can we just keep track of a `long previousRecordTimestamp` variable and any time we process a new window in the loop, we just check if `window.value.timestamp > previousRecordTimestamp` and if so then we set `previousRecordTimestamp` to this new value

Actually I think that sheds some light on all this: the whole "keep track of the latest left type window" idea was a clever way to find the previous record in the original design, before we knew  we could use ValueAndTimestamp to get the max  timestamp of a window. And we never went back over that clever hack and realized we could be less clever now, and just go with the simple approach of checking the max timestamp of every window

Reply:
Yeah I think this makes a lot of sense. The one catch I found is that we check to see if `latestLeftTypeWindow != null` when creating the current record's left window, and if we update `previousRecord` each time we see a window, we could get a number that isn't actually within range for the current record's left window. EX: record comes in at 30, timeDifference = 10. we have a window [10,20] but it's a right window, and within it the only record is at 12. since it's the max record within that window, we would store it in `previousRecordTimestamp` , but it's right window is [13,23] which we don't want to create. So I think storing just the long is fine (mostly typing it out to work out my thought process), and the check `rightWindowNecessaryAndPossible` would weed out us creating the [13,23] window on accident. Does this all seem right?

Reply:
Yeah, you're saying that we just always keep track of the `previousRecordTimestamp`, but before we go ahead and create a left window for the current window we just actually verify that the previous record is within range? That makes sense to me, actually if anything I feel like it will make `rightWindowNecessaryAndPossible` even more clear to put it in terms of `previousRecordTimestamp`. What I'm realizing from this is that it's easier to understand these boolean checks in terms of the actual record locations, in general. Maybe it's just my mental model, I still picture a rectangle sliding over boxes on a timeline 😄